### PR TITLE
feat(Trader): integration 1inch protocol 

### LINKED ISSUE #4015: [Bounty] Integrate 1inch, a dex aggregator, in Mask Plugin
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] 300 $MASK or $2000

## Part 1: Problem Description

**Background**
1inch is a dex aggregator on ETH BSC and Polygon. We have already integrated 0x on ETH but not on Polygon yet. We think 1inch will be a good fit for all chains

Previously defi plugins on Mask, like the trading widgets powered by 0x and Uniswap, has been plagued by high ETH gas prices.  However with the launch of polygon and BSC integration, the problem should have been largely resolved. Tradings app which was hindered by monstrous ETH gas price is now very much affordable. 

**Solution**
Create a plugin that integrate 1inch.  The UI should be identical to other swaps

polygon swaps
eth swaps

See the existing polygon swaps and ETH swaps.

Anylink cashtag with viable ticker, like $ETH on Twitter, could trigger the plugin. Swaps are also accessible via top line search bar and the swap entry on the left bar menu

Language like 
**I just swapped $aa worth of xxx token for yyy token, can I make money?** 
should be added to help spread the word.

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://1inch.io/
- https://docs.1inch.io/api/

## Part 3: Contact and Support
@yisiliu for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q
1inch discord server: https://discord.com/invite/hxPhPDW

**Submission Requirements**
Please include demo to show the integration is working.

**Winner Announcement Date**
Winners will be announced at the conclusion of Open DeFi Hackathon

**Judging Criteria**
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by developerfred**
closes #4015 

**Docs**
https://docs.1inch.io/api/quote-swap

**Using** ```createLookupTableResolver ```  https://github.com/DimensionDev/Maskbook/pull/4141

### CODE REVIEW

File: packages/maskbook/src/plugins/Trader/apis/1inch/index.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import type { NetworkType } from '@masknet/web3-shared'
+import urlcat from 'urlcat'
+
+export async function swapOneQuote(request: SwapQuoteOneRequest, networkType: NetworkType) {
+    const params: Record<string, string> = {}
+    Object.entries(request).map(([key, value]) => {
+        if (typeof value === 'string') params[key] = value
+    })
+    if (request.slippage) params.slippage = new BigNumber(request.slippage).dividedBy(BIPS_BASE).toFixed()
+    // if (request.buyTokenPercentageFee)
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/1inch/index.ts:20]: plz clear the commented code block

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 2]
```diff
@@ -108,6 +110,22 @@ export function resolveTradeProviderLink(tradeProvider: TradeProvider, networkTy
             return 'https://exchange.pancakeswap.finance/#/swap'
         case TradeProvider.DODO:
             return 'https://app.dodoex.io'
+        case TradeProvider.ONE_INCH:
+            switch (networkType) {
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/pipes.ts:114]: plz switch to table driven pattern

CC #4141 @guanbinrui

File: packages/maskbook/src/plugins/Trader/apis/1inch/index.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+export async function swapOneQuote(request: SwapQuoteOneRequest, networkType: NetworkType) {
+    const params: Record<string, string> = {}
+    Object.entries(request).map(([key, value]) => {
+        if (typeof value === 'string') params[key] = value
+    })
+    if (request.slippage) params.slippage = new BigNumber(request.slippage).dividedBy(BIPS_BASE).toFixed()
+
+    const response =
+        (await fetch(urlcat(ONE_INCH_BASE_URL[networkType], 'swap', params))) ||
+        (await fetch(urlcat(ONE_INCH_BASE_URL[networkType], 'quote', params)))
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/1inch/index.ts:23]: ??????

**Review by septs — CHANGES_REQUESTED** [review]
plz resolve build error

File: packages/maskbook/src/plugins/Trader/apis/1inch/index.ts (hunk 3 above)
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Trader/apis/1inch/index.ts:23]: Why are you're doing this? This doesn't make sense.

Comment (developerfred) [reply packages/maskbook/src/plugins/Trader/apis/1inch/index.ts]: I fixed, now i am using swap endpoint

**Comment by developerfred** [comment]
> plz resolve build error

for this build error to exit I need pr #4141 to be accepted
createLookupTableResolver #4141

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 4]
```diff
@@ -95,6 +96,22 @@ export function resolveTradeProviderLink(tradeProvider: TradeProvider, networkTy
...
+                case NetworkType.Ethereum:
+                    return 'https://api.1inch.exchange/v3.0/1/'
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/pipes.ts:102]:
```suggestion
                    return 'https://app.1inch.io/#/1/swap/'
```

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 5]
```diff
@@ -108,6 +125,8 @@ export function resolveTradePairLink(tradeProvider: TradeProvider, address: stri
             return `https://info.uniswap.org/pair/${address}`
         case TradeProvider.ZRX:
             return ''
+        case TradeProvider.ONE_INCH:
+            return ''
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/pipes.ts:129]: Can we use the pool link here?

https://app.1inch.io/#/1/dao/pools/0x6a11f3e5a01d129e566d783a7b6e8862bfd66cca/governance

File: packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        // start waiting for provider to confirm tx
+        setTradeState({
+            type: TransactionStateType.WAIT_FOR_CONFIRMING,
+        })
+
+        // estimate transaction
+        try {
+            await web3.eth.call(config)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts:53]: Please `web3.eth.estimateGas` here.

**Comment by guanbinrui** [comment]
@developerfred Please fix those build errors.

File: packages/maskbook/src/plugins/Trader/apis/1inch/index.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    const response_ = (await response.json()) as SwapQuoteOneResponse | SwapOneErrorResponse
+
+    const validationErrorResponse = response_ as SwapOneValidationErrorResponse
+    if (validationErrorResponse.code)
+        throw new Error(first(validationErrorResponse.validationErrors)?.reason ?? 'Unknown Error')
+
+    const serverErrorResponse = response_ as SwapOneServerErrorResponse
+    if (serverErrorResponse.reason)
+        throw new Error(first(validationErrorResponse.validationErrors)?.reason || 'Unknown Error')
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/apis/1inch/index.ts:31]:
```suggestion
        throw new Error(first(validationErrorResponse.validationErrors)?.reason ?? 'Unknown Error')
```

File: packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        // start waiting for provider to confirm tx
+        setTradeState({
+            type: TransactionStateType.WAIT_FOR_CONFIRMING,
+        })
+
+        // estimate transaction
+        try {
+            await web3.eth.estimateGas(config)
+        } catch {
+            // for some transactions will always fail if we do estimation before a kick to the chain
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts:55]: Please don't copy this here. Instead, it should implement its error handler.

File: packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                setTradeState({
+                    type: TransactionStateType.FAILED,
+                    error: new Error('User denied the transaction.'),
+                })
+                return
+            }
+        }
+
+        // compose transaction config
+        const config_ = {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts:70]: Not necessary.

File: packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts (hunk 8 above)
Comment (developerfred) [reply packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts:55]: Ok @guanbinrui, I will implement this.

File: packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts (hunk 9 above)
Comment (developerfred) [reply packages/maskbook/src/plugins/Trader/trader/1inch/useTradeCallback.ts:70]: ok, thanks!

**Comment by guanbinrui** [comment]
We need confirmation from  @DimensionDev/qa.

**Comment by developerfred** [comment]
> @developerfred Any updates?

I'm still working, I believe I can send a PR correcting it today.

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+import urlcat from 'urlcat'
+
+export async function swapOneQuote(request: SwapQuoteOneRequest, networkType: NetworkType) {
+    const params: Record<string, string | number> = {}
+    Object.entries(request).map(([key, value]) => {
+        params[key] = value
+    })
+    if (request.slippage) params.slippage = new BigNumber(request.slippage).dividedBy(BIPS_BASE).toFixed()
+
+    const response = await fetch(urlcat(ONE_INCH_BASE_URL[networkType], 'quote' || 'swap', params))
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]: 1. `'quote' || 'swap'` always returns `'quote'`, is it a mistake?
2. We have a `fetchJSON<T>`, try using it.

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    const response_ = (await response.json()) as SwapQuoteOneResponse | SwapOneErrorResponse
+
+    const validationErrorResponse = response_ as SwapOneValidationErrorResponse
+    if (validationErrorResponse.code)
+        throw new Error(first(validationErrorResponse.validationErrors)?.reason ?? 'Unknown Error')
+
+    const serverErrorResponse = response_ as SwapOneServerErrorResponse
+    if (serverErrorResponse.reason)
+        throw new Error(first(validationErrorResponse.validationErrors)?.reason || 'Unknown Error')
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]:
```suggestion
        throw new Error(first(validationErrorResponse.validationErrors)?.reason || 'Unknown Error')

    const serverErrorResponse = response_ as SwapOneServerErrorResponse
    if (serverErrorResponse.reason)
        throw new Error(first(validationErrorResponse.validationErrors)?.reason || 'Unknown Error')
```

File: packages/maskbook/src/plugins/Trader/apis/1inch/index.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+export async function swapOneQuote(request: SwapQuoteOneRequest, networkType: NetworkType) {
+    const params: Record<string, string> = {}
+    Object.entries(request).map(([key, value]) => {
+        if (typeof value === 'string') params[key] = value
+    })
+    if (request.slippage) params.slippage = new BigNumber(request.slippage).dividedBy(BIPS_BASE).toFixed()
+
+    const response =
+        (await fetch(urlcat(ONE_INCH_BASE_URL[networkType], 'swap', params))) ||
+        (await fetch(urlcat(ONE_INCH_BASE_URL[networkType], 'quote', params)))
```
Comment (UncleBill) [inline packages/maskbook/src/plugins/Trader/apis/1inch/index.ts]:
```suggestion
    const swapUrl = urlcat(ONE_INCH_BASE_URL[networkType], 'swap', params)
    const quoteUrl = urlcat(ONE_INCH_BASE_URL[networkType], 'quote', params)
    const response = (await fetch(swapUrl)) ||  (await fetch(quote))
```

Better to write code easy to understand within a glance.
We have so many dependency libraries that contain thousands lines of code, saving 1 or 2 lines of code doesn't really matter. Clean code would be better.

**Comment by guanbinrui** [comment]
Hi @developerfred Any updates?

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    const params: Record<string, string | number> = {}
+    Object.entries(request).map(([key, value]) => {
+        params[key] = value
+    })
+    if (request.slippage) params.slippage = new BigNumber(request.slippage).dividedBy(BIPS_BASE).toFixed()
+
+    const netType: NetworkType = getNetworkTypeFromChainId(request.chainId)!
+
+    const response = await fetch(urlcat(ONE_INCH_BASE_URL[netType], 'swap', params))
+    const response_ = (await response.json()) as SwapQuoteOneResponse | SwapOneErrorResponse
```
Comment (septs) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]:
```suggestion
    const response_ = await fetchJSON<SwapQuoteOneResponse | SwapOneErrorResponse>(urlcat(ONE_INCH_BASE_URL[netType], 'swap', params))
```

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+    const validationErrorResponse = response_ as SwapOneValidationErrorResponse
+    if (validationErrorResponse.code)
+        throw new Error(first(validationErrorResponse.validationErrors)?.reason ?? 'Unknown Error')
```
Comment (septs) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]:
```suggestion
    if ('code' in response_)
        throw new Error(first(response_.validationErrors)?.reason ?? 'Unknown Error')
```

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+    const serverErrorResponse = response_ as SwapOneServerErrorResponse
+    if (serverErrorResponse.reason)
+        throw new Error(first(validationErrorResponse.validationErrors)?.reason || 'Unknown Error')
```
Comment (septs) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]:
```suggestion
    if ('reason' in response_)
        throw new Error(first(response_.validationErrors)?.reason || 'Unknown Error')
```

**Comment by developerfred** [comment]
done @septs

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+import urlcat from 'urlcat'
+
+import { getNetworkTypeFromChainId, NetworkType } from '@masknet/web3-shared-evm'
+import { fetchJSON } from '@masknet/web3-providers'
+
+export async function swapOneQuote(request: SwapQuoteOneRequest) {
+    const params: Record<string, string | number> = {}
+    Object.entries(request).map(([key, value]) => {
+        params[key] = value
+    })
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]:
```suggestion
    const params: Record<string, string | number> =  Object.fromEntries(Object.entries(request))
```

File: packages/web3-constants/evm/rpc.json [hunk 17]
```diff
@@ -59,8 +59,7 @@
         "Celo": ["https://forno.celo.org"],
         "Fantom": ["https://rpc.ftm.tools/"],
         "Aurora": ["https://mainnet.aurora.dev"],
-        "Aurora_Testnet": ["https://testnet.aurora.dev"],
-        "Conflux": ["https://evm.confluxrpc.com"]
```
Comment (UncleBill) [inline packages/web3-constants/evm/rpc.json]: Do not remove configs for other chains.

File: packages/web3-constants/evm/token-list.json [hunk 18]
```diff
@@ -17,7 +17,6 @@
         "Celo": ["https://tokens.r2d2.to/latest/42220/tokens.json"],
         "Fantom": ["https://tokens.r2d2.to/latest/250/tokens.json"],
         "Aurora": ["https://tokens.r2d2.to/latest/1313161554/tokens.json"],
-        "Aurora_Testnet": [],
-        "Conflux": ["https://tokens.r2d2.to/latest/1030/tokens.json"]
```
Comment (UncleBill) [inline packages/web3-constants/evm/token-list.json]: ditto

File: packages/web3-constants/evm/zerion.json [hunk 19]
```diff
@@ -17,8 +17,7 @@
         "Celo": "celo-assets",
         "Fantom": "fantom-assets",
         "Aurora": "aurora-assets",
-        "Aurora_Testnet": "",
-        "Conflux": ""
```
Comment (UncleBill) [inline packages/web3-constants/evm/zerion.json]: ditto

File: packages/web3-constants/evm/token.json [hunk 20]
```diff
@@ -17,8 +17,7 @@
         "Celo": "0x471EcE3750Da237f93B8E339c536989b8978a438",
         "Fantom": "0x21be370d5312f44cb42ce377bc9b8a0cef1a4c83",
         "Aurora": "0xC9BdeEd33CD01541e1eeD10f90519d2C06Fe3feB",
-        "Aurora_Testnet": "",
-        "Conflux": ""
```
Comment (UncleBill) [inline packages/web3-constants/evm/token.json]: ditto

Please do double check.

**Review by UncleBill — CHANGES_REQUESTED** [review]

File: packages/web3-constants/evm/rpc.json (hunk 17 above)
Comment (developerfred) [reply packages/web3-constants/evm/rpc.json]: @UncleBill conflux pull request did not add "Conflux" in compile constants, when you make a commit and the interpreter deletes what is not defined in compile_constants

File: packages/web3-constants/evm/ito.json [hunk 21]
```diff
@@ -58,7 +60,7 @@
         "Fantom": "0x981be454a930479d92C91a0092D204b64845A5D6",
         "Aurora": "0x2cf91AD8C175305EBe6970Bd8f81231585EFbd77",
         "Aurora_Testnet": "0xdcA6F476EebCDE8FE8b072e3fC80dBC28dC209b3",
-        "Conflux": "0x066804d9123bf2609ed4a4a40b1177a9c5a9ed51"
+        "Conflux": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/ito.json]: Plz revert it.

File: packages/web3-constants/evm/ito.json [hunk 22]
```diff
@@ -79,7 +81,7 @@
         "Fantom": 25071597,
         "Aurora": 57350598,
         "Aurora_Testnet": 77919102,
-        "Conflux": 37722805
+        "Conflux": 0
```
Comment (guanbinrui) [inline packages/web3-constants/evm/ito.json]: Plz revert it.

File: packages/web3-constants/evm/ito.json [hunk 23]
```diff
@@ -120,7 +123,7 @@
         "Fantom": "0x83D6b366f21e413f214EB077D5378478e71a5eD2",
         "Aurora": "0x578a7Fee5f0D8CEc7d00578Bf37374C5b95C4b98",
         "Aurora_Testnet": "",
-        "Conflux": "0x05ee315e407c21a594f807d61d6cc11306d1f149"
+        "Conflux": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/ito.json]: Plz revert it.

File: packages/web3-constants/evm/nft-red-packet.json [hunk 24]
```diff
@@ -18,7 +18,7 @@
         "Fantom": "",
         "Aurora": "0x05ee315E407C21a594f807D61d6CC11306D1F149",
         "Aurora_Testnet": "0x97369fEE7db34E0BfE47861f2ec44b4378d13eB4",
-        "Conflux": "0x5b966f3a32db9c180843bcb40267a66b73e4f022"
+        "Conflux": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/nft-red-packet.json]: Plz revert it.

File: packages/web3-constants/evm/rpc.json [hunk 25]
```diff
@@ -81,6 +81,6 @@
         "Fantom": [0, 0, 0, 0, 0],
         "Aurora": [0, 0, 0, 0, 0],
         "Aurora_Testnet": [0, 0, 0, 0, 0],
-        "Conflux": [0, 0, 0, 0, 0]
+        "Conflux": []
```
Comment (guanbinrui) [inline packages/web3-constants/evm/rpc.json]: Plz revert it.

File: packages/web3-constants/evm/red-packet.json [hunk 26]
```diff
@@ -99,7 +102,7 @@
         "Fantom": 25112473,
         "Aurora": 57552338,
         "Aurora_Testnet": 77918765,
-        "Conflux": 37670572
+        "Conflux": 0
```
Comment (guanbinrui) [inline packages/web3-constants/evm/red-packet.json]: Plz revert it.

File: packages/web3-constants/evm/red-packet.json [hunk 27]
```diff
@@ -78,7 +81,7 @@
         "Fantom": "0x578a7Fee5f0D8CEc7d00578Bf37374C5b95C4b98",
         "Aurora": "0x19f179D7e0D7d9F9d5386afFF64271D98A91615B",
         "Aurora_Testnet": "0xdB93cCd481012bB5D1E2c8d0aF7C5f2940c00fdC",
-        "Conflux": "0x96c7d011cdfd467f551605f0f5fce279f86f4186"
+        "Conflux": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/red-packet.json]: Plz revert it.

File: packages/web3-constants/evm/token-list.json [hunk 28]
```diff
@@ -18,6 +18,6 @@
         "Fantom": ["https://tokens.r2d2.to/latest/250/tokens.json"],
         "Aurora": ["https://tokens.r2d2.to/latest/1313161554/tokens.json"],
         "Aurora_Testnet": [],
-        "Conflux": ["https://tokens.r2d2.to/latest/1030/tokens.json"]
+        "Conflux": []
```
Comment (guanbinrui) [inline packages/web3-constants/evm/token-list.json]: Plz revert it.

File: packages/web3-constants/evm/token.json [hunk 29]
```diff
@@ -984,6 +984,6 @@
         "Fantom": "0x0000000000000000000000000000000000000000",
         "Aurora": "0x0000000000000000000000000000000000000000",
         "Aurora_Testnet": "",
-        "Conflux": "0x0000000000000000000000000000000000000000"
+        "Conflux": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/token.json]: Plz revert it.

**Comment by developerfred** [comment]
Revert done @guanbinrui

File: packages/maskbook/src/plugins/Trader/trader/1inch/useTrade.ts [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+} from '@masknet/web3-shared-evm'
+import { useAsyncRetry } from 'react-use'
+import { safeUnreachable } from '@dimensiondev/kit'
+import { PluginTraderRPC } from '../../messages'
+import { TradeStrategy } from '../../types'
+import { useSlippageTolerance } from '../0x/useSlippageTolerance'
+import { useTradeProviderSettings } from '../useTradeSettings'
+import { currentNetworkSettings } from '../../../Wallet/settings'
+
+export function setTokenNativeNetwork(networkType: NetworkType) {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/trader/1inch/useTrade.ts]:
```suggestion
export function resolveTokenNativeNetwork(networkType: NetworkType) {
```

File: packages/public-api/src/web.ts [hunk 31]
```diff
@@ -193,6 +193,7 @@ export enum TradeProvider {
     PANCAKESWAP = 6,
     DODO = 7,
     UNISWAP_V3 = 8,
+    ONE_INCH = 10,
```
Comment (guanbinrui) [inline packages/public-api/src/web.ts]: Duplicated.

File: packages/maskbook/src/plugins/Trader/apis/1inch/index.ts [hunk 32]
```diff
@@ -0,0 +1,37 @@
+import BigNumber from 'bignumber.js'
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Trader/apis/1inch/index.ts:1]: We don't have a `maskbook/` folder.

File: packages/web3-constants/evm/aave.json [hunk 33]
```diff
@@ -17,7 +17,8 @@
         "Celo": "",
         "Fantom": "",
         "Aurora": "",
-        "Aurora_Testnet": ""
+        "Aurora_Testnet": "",
+        "Conflux": ""
```
Comment (guanbinrui) [inline packages/web3-constants/evm/aave.json]: Strange. Why did Conflux changes include here? Plz merge the mainstream.

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 34]
```diff
@@ -0,0 +1,27 @@
...
+
+    const netType: NetworkType = getNetworkTypeFromChainId(request.chainId)!
+
+    const response_ = await fetchJSON<SwapQuoteOneResponse | SwapOneErrorResponse>(
+        urlcat(ONE_INCH_BASE_URL[netType], 'swap', params),
+    )
+
+    if ('code' in response_) throw new Error(first(response_.validationErrors)?.reason ?? 'Unknown Error')
+
+    if ('reason' in response_) throw new Error(first(response_.validationErrors)?.reason || 'Unknown Error')
```
Comment (septs) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]:
```suggestion
    if ('code' in response_ || 'reason' in response_) {
        throw new Error(first(response_.validationErrors)?.reason ?? 'Unknown Error')
    }
```

File: packages/maskbook/src/plugins/Trader/trader/1inch/useTrade.ts [hunk 35]
```diff
@@ -0,0 +1,82 @@
...
+        case NetworkType.Ethereum:
+            return 'ETH'
+        case NetworkType.Binance:
+            return '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
+        case NetworkType.Polygon:
+            return '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
+        case NetworkType.Arbitrum:
+            return '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
+        case NetworkType.xDai:
+            return '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/trader/1inch/useTrade.ts]:
```suggestion
        case NetworkType.Binance:
        case NetworkType.Polygon:
        case NetworkType.Arbitrum:
        case NetworkType.xDai:
            return '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
```

File: packages/maskbook/src/plugins/Trader/apis/1inch/index.ts (hunk 32 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Trader/apis/1inch/index.ts:1]: @developerfred Did you use the wrong file path?

Comment (developerfred) [reply packages/maskbook/src/plugins/Trader/apis/1inch/index.ts:1]: I believe this branch was outdated so it got the wrong folder path

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 36]
```diff
@@ -0,0 +1,27 @@
...
+import { first } from 'lodash-unified'
+import { BIPS_BASE, ONE_INCH_BASE_URL } from '../../constants'
+import type { SwapOneErrorResponse, SwapQuoteOneResponse, SwapQuoteOneRequest } from '../../types'
+import urlcat from 'urlcat'
+
+import { getNetworkTypeFromChainId, NetworkType } from '@masknet/web3-shared-evm'
+import { fetchJSON } from '../../../../../../web3-providers/src/helpers'
+
+export async function swapOneQuote(request: SwapQuoteOneRequest) {
+    const params: Record<string, string | number> = Object.fromEntries(Object.entries(request))
```
Comment (Jack-Works) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]: why

**Review by guanbinrui — DISMISSED** [review]

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/mask/src/plugins/Trader/trader/1inch/useTradeCallback.ts [hunk 37]
```diff
@@ -0,0 +1,97 @@
...
+        type: TransactionStateType.UNKNOWN,
+    })
+
+    // compose transaction config
+    const config = useMemo(() => {
+        if (!account || !tradeComputed?.trade_ || ![ChainId.Mainnet, ChainId.BSC, ChainId.Matic].includes(chainId))
+            return null
+        return {
+            from: account,
+            ...pick(tradeComputed.trade_, ['to', 'data', 'value', 'gas', 'gasPrice']),
```
Comment (Jack-Works) [inline packages/mask/src/plugins/Trader/trader/1inch/useTradeCallback.ts:33]: Do not use `pick`. It's not type safe.

File: packages/mask/src/plugins/Trader/trader/1inch/useTradeGasLimit.ts [hunk 38]
```diff
@@ -0,0 +1,27 @@
...
+export function useTradeGasLimit(tradeComputed: TradeComputed<SwapQuoteOneResponse> | null): AsyncState<number> {
+    const { targetChainId } = TargetChainIdContext.useContainer()
+
+    const web3 = useWeb3({ chainId: targetChainId })
+    const account = useAccount()
+    const config = useMemo(() => {
+        if (!account || !tradeComputed?.trade_) return null
+        return {
+            from: account,
+            ...pick(tradeComputed.trade_, ['to', 'data', 'value']),
```
Comment (Jack-Works) [inline packages/mask/src/plugins/Trader/trader/1inch/useTradeGasLimit.ts:19]: Do not use pick. It's not type safe.

File: packages/maskbook/src/plugins/Trader/apis/1inch/index.ts (hunk 32 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/Trader/apis/1inch/index.ts:1]: ping

**Comment by wenluomask** [comment]
Hi @developerfred ,

Please make a new build in CI and merge 2.8.0(the latest production version) into this pr.

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts (hunk 36 above)
Comment (UncleBill) [reply packages/mask/src/plugins/Trader/apis/1inch/index.ts]: @developerfred You can do shallow clone with
```suggestion
    const params = { ...request }
```

But I don't think you really need cloning. Following line should not mix in this function, it should be moved outside, that means you should make the the parameters satisfy this function instead.

```ts
    if (request.slippage) params.slippage = new BigNumber(request.slippage).dividedBy(BIPS_BASE).toFixed()
```

File: packages/mask/src/plugins/Trader/apis/1inch/index.ts [hunk 39]
```diff
@@ -0,0 +1,27 @@
...
+
+    if (request.slippage) params.slippage = new BigNumber(request.slippage).dividedBy(BIPS_BASE).toFixed()
+
+    const netType: NetworkType = getNetworkTypeFromChainId(request.chainId)!
+
+    const response_ = await fetchJSON<SwapQuoteOneResponse | SwapOneErrorResponse>(
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/apis/1inch/index.ts]:
```suggestion
    const response = await fetchJSON<SwapQuoteOneResponse | SwapOneErrorResponse>(
```

File: packages/mask/src/plugins/Trader/types/1inch.ts [hunk 40]
```diff
@@ -0,0 +1,77 @@
+import type { FungibleTokenDetailed } from '@masknet/web3-shared-evm'
+
+// Learn more from https://docs.1inch.io/api/quote-swap#swap
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/types/1inch.ts]: https://docs.1inch.io/api/quote-swap#swap 404, not found

File: packages/mask/src/plugins/Trader/types/1inch.ts [hunk 41]
```diff
@@ -0,0 +1,77 @@
...
+    gasPrice: string
+    gas: number // rough estimate, increase gas value by ~25%
+}
+export interface OneInchProtocols {
+    name: string
+    part: number
+    fromTokenAddress: string
+    toTokenAddress: string
+}
+// more information https://docs.1inch.io/api/quote-swap#description-of-query-parameters-1
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/types/1inch.ts]: ditto.

File: packages/web3-constants/evm/coingecko.json [hunk 42]
```diff
@@ -39,6 +39,6 @@
         "Fantom": "fantom",
         "Aurora": "ethereum",
         "Aurora_Testnet": "",
-        "Conflux": "conflux-token"
+        "Conflux": ""
```
Comment (UncleBill) [inline packages/web3-constants/evm/coingecko.json]: Please do self checking.

**Review by UncleBill — CHANGES_REQUESTED** [review]

File: packages/mask/src/plugins/Trader/trader/1inch/useTradeGasLimit.ts (hunk 38 above)
Comment (developerfred) [reply packages/mask/src/plugins/Trader/trader/1inch/useTradeGasLimit.ts:19]: @Jack-Works what is the safe typo in this case?

Comment (developerfred) [reply packages/mask/src/plugins/Trader/trader/1inch/useTradeGasLimit.ts:19]: pickBy?

**Pull request closed without merging**